django/core/servers/basehttp.py

```diff
@@ -81,6 +82,29 @@ class ThreadedWSGIServer(socketserver.ThreadingMixIn, WSGIServer):
     """A threaded version of the WSGIServer"""
     daemon_threads = True
 
+    def __init__(self, *args, connections_override=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.connections_override = connections_override
+
+    # socketserver.ThreadingMixIn.process_request() passes this method as
+    # the target to a new Thread object.
+    def process_request_thread(self, request, client_address):
+        if self.connections_override:
+            # Override this thread's database connections with the ones
+            # provided by the parent thread.
+            for alias, conn in self.connections_override.items():
+                connections[alias] = conn
+        super().process_request_thread(request, client_address)
```

Comment:
Do we need this methods? Maybe I'm missing sth but thread's database connections should already by overridden, and we don't pass `connections_override` to the `_create_server()` anywhere in Django :thinking:

Reply:
It looks that we're adding a lot of logic only for tests. Am I right?

Reply:
Thanks a lot for taking a look at this patch, @felixxm.

For this PR, I used the same testing strategy I used when fixing the analogous bug in `LiveServerTestCase` (https://github.com/django/django/commit/f6cd669ff203192c29495174e53da6b16883b039) that preceded the bug being fixed here.

That strategy was to use `LiveServerTestCase._create_server_thread(connections_override)` to pass in an explicit connection during the test so the test has a specific connection against which it can test is closed. At that time, the ability to pass in `connections_override` already existed to support reusing the same connection when using an in-memory sqlite database. That ability was added when `LiveServerTestCase` was first introduced: https://github.com/django/django/commit/2f02a05ffb45be68b4164b4785ff1826833150a3

With respect to the current PR, `ThreadedWSGIServer` doesn't yet support the ability to pass in `connections_override`. This is why `ThreadedWSGIServer` doesn't fully support being used with an in-memory sqlite database. That is the subject of ticket [#29062](https://code.djangoproject.com/ticket/29062) and is why I needed to add that ability here. Thus, you can view this PR is doing the work needed to fix two fix two bugs, which is why it seems a lot is being changed here. Merging this PR will allow that bug to be fixed easily. I commented to that effect on that ticket here: https://code.djangoproject.com/ticket/29062#comment:18

The `ThreadedWSGIServer._close_connections()` method is the only method that was added for the test that wouldn't already be needed to fix ticket 29062 (hence the code comment "This is exposed separately so it can be overridden in tests"). Its purpose is to provide a hook so tests can know when `connections.close_all()` has been called. It lets the test insert a threading event so the test knows immediately when it can check that the connection has been closed. Without such a hook, the test would need to do something like insert an arbitrary delay using `time.sleep(0.25)`, which would (1) make the test slower, and (2) introduce a source of test flakiness. The hook and event approach solves both (1) and (2), but at the expense of adding a method.

Reply:
PS - you might ask: why not fix ticket 29062 first? The reason is that fixing ticket 29062 properly would involve properly closing connections. Thus, any correct fix of ticket 29062 would be a superset of this PR, which would make it an even bigger change.

Reply:
> PS - you might ask: why not fix ticket 29062 first? The reason is that fixing ticket 29062 properly would involve properly closing connections. Thus, any correct fix of ticket 29062 would be a superset of this PR, which would make it an even bigger change.

Yes that was my first question :smile: Thanks for details :+1: . I'm afraid that it can be still confusing for a future me. I will try to move something to a separate commit e.g. `_make_connections_override()` which should reduce the number of changes and make it easier to bisect and fix potential regression.

Reply:
> I will try to move something to a separate commit 

Okay, sure. Thanks, @felixxm. By the way, you might also want to read this comment of mine, which precedes the one I linked to above: https://code.djangoproject.com/ticket/29062#comment:18
It relates to code you would be separating out into a separate commit.

Reply:
> `_make_connections_override()`

Yes, that's a good one (a simple refactor) to separate out into a separate commit.